logic/LogicManager.java

```diff
@@ -106,12 +107,21 @@ private FunctionalIterator<Rule> rulesWithNegations() {
         return rules().filter(rule -> !rule.when().negations().isEmpty());
     }
 
-    public FunctionalIterator<Unifier> unifiers(Concludable concludable, Rule rule) {
-        return iterate(applicableRules(concludable).getOrDefault(rule, new HashSet<>()));
+    public Map<Rule, Set<Unifier>> applicableRules(Concludable concludable) {
+        Map<Rule, Set<Unifier>> unifiers = logicCache.unifiers().getIfPresent(concludable);
+        if (unifiers == null) {
+            unifiers = concludable.computeApplicableRules(conceptMgr, this);
+        }
+        return unifiers;
```

Comment:
Uses the cache if the unifiers are present . Does not add to the cache if they are not.